flocker/ca/test/test_ca.py

```diff
@@ -16,13 +16,296 @@
 from twisted.python.filepath import FilePath
 
 from .. import (RootCredential, ControlCredential, NodeCredential,
-                PathError, EXPIRY_20_YEARS, AUTHORITY_CERTIFICATE_FILENAME,
-                AUTHORITY_KEY_FILENAME, CONTROL_CERTIFICATE_FILENAME,
-                CONTROL_KEY_FILENAME)
+                UserCredential, PathError, EXPIRY_20_YEARS,
+                AUTHORITY_CERTIFICATE_FILENAME, AUTHORITY_KEY_FILENAME,
+                CONTROL_CERTIFICATE_FILENAME, CONTROL_KEY_FILENAME)
 
 from ...testtools import not_root, skip_on_broken_permissions
 
 
+class UserCredentialTests(SynchronousTestCase):
+    """
+    Tests for ``flocker.ca._ca.UserCredential``.
+    """
+    def setUp(self):
+        """
+        Generate a RootCredential for the API certificate tests
+        to work with.
+        """
+        self.path = FilePath(self.mktemp())
+        self.path.makedirs()
+        self.ca = RootCredential.initialize(self.path, b"mycluster")
+        self.username = b"alice"
+
+    def test_written_keypair_exists(self):
+        """
+        ``UserCredential.initialize`` writes a PEM file to the
+        specified path.
+        """
+        uc = UserCredential.initialize(self.path, self.ca, self.username)
+        cert_file = b"{user}.crt".format(user=uc.username)
+        key_file = b"{user}.key".format(user=uc.username)
+        self.assertEqual(
+            (True, True),
+            (self.path.child(cert_file).exists(),
+             self.path.child(key_file).exists())
+        )
+
+    def test_decoded_certificate_matches_public_key(self):
+        """
+        A decoded certificate's public key matches the public key it is
+        meant to be paired with.
+        """
+        uc = UserCredential.initialize(self.path, self.ca, self.username)
+        self.assertTrue(
+            uc.credential.keypair.keypair.matches(
+                uc.credential.certificate.getPublicKey())
+        )
+
+    def test_decoded_certificate_matches_private_key(self):
+        """
+        A decoded certificate matches the private key it is meant to
+        be paired with.
+        """
+        uc = UserCredential.initialize(self.path, self.ca, self.username)
+        priv = uc.credential.keypair.keypair.original
+        pub = uc.credential.certificate.getPublicKey().original
+        pub_asn1 = crypto.dump_privatekey(crypto.FILETYPE_ASN1, pub)
+        priv_asn1 = crypto.dump_privatekey(crypto.FILETYPE_ASN1, priv)
+        pub_der = asn1.DerSequence()
+        pub_der.decode(pub_asn1)
+        priv_der = asn1.DerSequence()
+        priv_der.decode(priv_asn1)
+        pub_modulus = pub_der[1]
+        priv_modulus = priv_der[1]
+        self.assertEqual(pub_modulus, priv_modulus)
```

Comment:
`OpenSSL.Context.check_privatekey` (pyOpenSSL 0.15) will do this comparison for you.